### ISSUE TITLE
Wrong interaction *print-circle* *PRINT-READABLY* and vectors

### ISSUE THREAD

**Issue by kpoeck** [clasp-developers/clasp#706]
```lisp
(let ((vector (vector 23))
               (*PRINT-READABLY* T)
               (*PRINT-CIRCLE* t)
               (*PRINT-PRETTY* T)
               )
           (write-to-string vector))
->
"#A#1=(T (1) #1#)"
````
but there is no circle

Works fine with *PRINT-READABLY* nil
```lisp
(let ((vector (vector 23 24 25))
               (*PRINT-READABLY* nil)
               (*PRINT-CIRCLE* t)
               (*PRINT-PRETTY* T)
               )
           (write-to-string vector))
"#(23 24 25)"
````
and even real circles are detected fine
```lisp
(let ((vector (vector 25 26 27))
      (*PRINT-READABLY* nil)
      (*PRINT-CIRCLE* t)
      (*PRINT-PRETTY* T)
      )
  (setf (aref vector 2) vector)
  (write-to-string vector))
"#1=#(25 26 #1#)"
````

**Comment by kpoeck** [comment]
Must be somewhere around
do-pprint-logical-block
but haven't be able to understand the problem. Tracing prints again, so might introduce Heisenbugs.

perhaps *print-readable* prints the vector just another time, so everything is considered a circle

**Comment by kpoeck** [comment]
Look also at
```lisp
(let ((array (make-array (list 2 2) :initial-contents '((5 6)(7 8))))
              (*PRINT-READABLY* T)
              (*PRINT-CIRCLE* t)
              (*PRINT-PRETTY* T)
              )
          (write-to-string array))
-> 
"#A#1=(T (2 2) #1#)"
````

**Comment by kpoeck** [comment]
This is caused by the change in do-pprint-logical-block
```lisp
(and (symbolp object) (symbol-package object)
                   ;; This is so that we do handle print-circle even when e.g. pretty printing a vector
                   ;; (for which object = nil here).
                   ;; It doesn't SEEM to result in (NIL . NIL) being printed as (#1=NIL . #1#), but
                   ;; I don't understand this very well.
                   (not (null object)))
````

* If I drop the (not (null object)) my examples w/o circularity print fine (good)
* But circularity in vectors & arrays -> stack-overflow or clasp dies in slime (real bad)
* ecl has nearly the same code in pprint.lsp (apart from the (not (null object)) and the same behaviour (prints fine w/o circularity), stack-overflow with circularity

**Comment by kpoeck** [comment]
w/o *print-pretty* nil all works fine, even with circles

**Comment by kpoeck** [comment]
I don't understand the code-change at all, seems that (search-print-circle object) is here always called with nil and the circle is not in nil

**Comment by Bike** [comment]
I changed it related to #68. But as the comment says I'm not confident I understand the pretty printer.

**Comment by kpoeck** [comment]
Rediscovered the problem with 
```lisp
(with-standard-io-syntax
  (let ((*PRINT-CIRCLE* T)
        (*PRINT-READABLY* T)
        (*PRINT-PRETTY* T))
    (with-output-to-string (s)
      (write #2A((3 4 5)) :stream s))))
->
"#A#1=(T (1 3) #1#)"
````

**Comment by Bike** [comment]
Okay so I guess the problem is that `do-pprint-logical-block` does circularity checks on the object it's provided. We frequently use the idiom of using `nil` as the object and then printing something unrelated. In this case we do that while printing the array contents (in `pprint-raw-array`), so `nil` gets registered as neading circularity. But then when we go to actually pprint `nil` itself it uses the circular reference. At least that's my impression from `(let ((*print-readably* t) (*print-circle* T)) (core::pprint-array #()))`.

In other words, we want `#A(T (0) ())`, for example. When we start printing the `(t (0) ())`, we do a `pprint-logical-block` with object `nil`. Then within that we want to actually print `nil`, so suddenly it's a circular reference.

What I'm thinking is that my change to add the `(not (null object))` was not the right way to fix the infinite depth problem. Dunno what is. Staring at it now

**Comment by Bike** [comment]
or i guess i could just merge the PR, but it's marked WIP still

**Comment by kpoeck** [comment]
Let me look at the pr on the weekend, fails a test, thats why I put it wip

**Comment by Bike** [comment]
I think this was fixed by #1163

**Comment by kpoeck** [comment]
right
